### PR TITLE
Port tagged-literals from ClojureScript

### PR DESCRIPTION
**Pull request by djblue**

### CODE REVIEW

File: clj/src/cljd/core.cljd [hunk 1]
```diff
@@ -7162,7 +7162,48 @@
...
+
+  IEquiv
+  (-equiv [this other]
+    (and (dart/is? other TaggedLiteral)
+         (= tag (.-tag other))
+         (= form (.-form other))))
+
+  IHash
+  (-hash [this]
+    (+ (* 31 (hash tag))
```
Comment (cgrand) [inline clj/src/cljd/core.cljd:7178]: use `u32-add` instead of `+`

File: clj/src/cljd/core.cljd [hunk 2]
```diff
@@ -7194,7 +7194,49 @@
...
-(def tagged-literal nil) ; TODO hack for macros but agged-literal will have to be implemented in cljd
+(deftype TaggedLiteral [tag form]
+  Object
+  (toString [coll]
+    (pr-str coll))
+
+  IEquiv
+  (-equiv [this other]
+    (and (dart/is? other TaggedLiteral)
+         (= tag (.-tag other))
```
Comment (djblue) [inline clj/src/cljd/core.cljd:7205]: Should `other` here be typed hinted as a `TaggedLiteral`?

Comment (dupuchba) [reply clj/src/cljd/core.cljd:7205]: yes it should

File: clj/src/cljd/core.cljd [hunk 3]
```diff
@@ -7194,7 +7194,49 @@
...
+         (= form (.-form other))))
+
+  IHash
+  (-hash [this]
+    (u32-add
```
Comment (dupuchba) [inline clj/src/cljd/core.cljd:7210]: here you should cache hashing with ensure hash

File: clj/src/cljd/core.cljd [hunk 4]
```diff
@@ -7194,7 +7194,49 @@
...
+     (* 31 (hash tag))
+     (hash form)))
+  (-hash-realized? [o] true)
```
Comment (dupuchba) [inline clj/src/cljd/core.cljd:7213]: it's realised if there is a value != -1 in __hash

File: clj/src/cljd/core.cljd (hunk 2 above)
Comment (dupuchba) [reply clj/src/cljd/core.cljd:7205]: e.g: `(.-tag ^TaggedLiteral other)`

File: clj/src/cljd/core.cljd [hunk 5]
```diff
@@ -7194,7 +7194,49 @@
...
+  "Return true if the value is the data representation of a tagged literal"
+  [value]
+  (dart/is? value TaggedLiteral))
+
+(defn tagged-literal
+  "Construct a data representation of a tagged literal from a
+  tag symbol and a form."
+  [tag form]
+  {:pre [(symbol? tag)]}
+  (TaggedLiteral. tag form))
```
Comment (dupuchba) [inline clj/src/cljd/core.cljd:7239]: `(TaggedLiteral. tag form -1)`

**Review by dupuchba — CHANGES_REQUESTED** [review]
thanks a lot, good to go after changes I asked

**Pull request merged**